### ISSUE TITLE
Seeking not accurate on FLAC files with WMFDecoder

### ISSUE THREAD

**Issue by fladd** [pyglet/pyglet#398]
On Windows, it seems that FLAC files are opened with WMFDEcoder instead of FFMPEG for some reason.
Unfortunately with this decoder, seeking is not accurate. For instance, if a FLAC file is 336 seconds long, and I seek to 335, then the audio will still play for several seconds, hence exceeding the actual playtime of the song (and indicating that it actually seeked to a different point in the audio file). Notably, the reported time by `player.time` has the same inaccuracy and will also report times longer than the actual runtime of the file. 

On MacOS, where FLAC files are decoded with FFMPEG, the problem is not present.

How can I make pyglet on Windows not use WMFDecoder, but FFMPEG for FLAC files instead?

**Comment by benmoran56** [comment]
The short answer is that you can manually make an instance of any Decoder, and then pass it in when loading:
```py
src = media.load(...., decoder=decoder_instance)
```
The doc string here has more information:
https://github.com/pyglet/pyglet/blob/e6372c9312970b032fcf93266381fc11f66bb79d/pyglet/media/__init__.py#L102


The long answer is that streaming decoding doesn't guarantee accuracy with seeking. As you've seen some decoders are more accurate than others, but if you need accurate seeking you should use StaticSources. Passing `streaming=False` in while loading will do this. 

The WMFDecoder is available on newer consumer versions of Windows, and allows playback of some popular compressed audio formats without needing FFmpeg.

**Comment by fladd** [comment]
Thanks for the explanation @benmoran56 , but could you maybe give me a more specific pointer to how this needs to be done? I assumed I could just do:

```
pyglet.media.load(track.name, file=track, decoder=pyglet.media.codecs.ffmpeg.FFmpegDecoder()))
```

But this leads to the following error:
```
  File "C:\Users\flokra\Documents\GitHub\ZAP\env\lib\site-packages\pyglet\media\__init__.py", line 125, in load
    return decoder.decode(file, filename, streaming)
  File "C:\Users\flokra\Documents\GitHub\ZAP\env\lib\site-packages\pyglet\media\codecs\ffmpeg.py", line 1050, in decode
    return FFmpegSource(filename, file)
  File "C:\Users\flokra\Documents\GitHub\ZAP\env\lib\site-packages\pyglet\media\codecs\ffmpeg.py", line 468, in __init__
    self._file = ffmpeg_open_filename(asbytes_filename(filename))
  File "C:\Users\flokra\Documents\GitHub\ZAP\env\lib\site-packages\pyglet\media\codecs\ffmpeg.py", line 140, in ffmpeg_open_filename
    raise FFmpegException('avformat_open_input in ffmpeg_open_filename returned an error opening file '
pyglet.media.codecs.ffmpeg.FFmpegException: avformat_open_input in ffmpeg_open_filename returned an error opening file C:\Users\flokra\AppDat
a\Local\Temp\tmps02ltli0 Error code: -13
```

**Comment by fladd** [comment]
Yeah, so this seems to be a Pyglet problem. To be more precise, `pyglet.media.codecs.ffmpeg.py` in line 464 creates a `NamedTemproraryFile`. The file is then left open, and the file**name** is then given (via  `asbytes_filename`) to `ffmpeg_open_filename` in line 468, which will try to open and read from that file again. According to the official Python documentation, this does not work on Windows: "Whether the name can be used to open the file a second time, while the named temporary file is still open, varies across platforms (it can be so used on Unix; it cannot on Windows NT or later)." (https://docs.python.org/3.8/library/tempfile.html#tempfile.NamedTemporaryFile). The solution is to instantiate `NamedTemporaryFile` with the `delete=False` parameter in line 464, and then close it after writing to it (and delete the file after closing it in the `__del__` method).

**Comment by fladd** [comment]
I reported this as an additional issue (#400) and created a pull request that should fix it (#401).

**Comment by caffeinepills** [comment]
This is a 2 bug report. Seek accuracy and opening memory files with ffmpeg.

As stated, the accuracy varies simply depending on how big of the buffers returned by the decoder and what is requested by the driver. You could try forcing a DirectSound audio driver for Windows to see if you get better seeking accuracy.

As far as ffmpeg memory loading, I did get his working, but the accuracy is just as inaccurate. But if you need the functionality you can try messing with my branch: https://github.com/caffeinepills/pyglet/tree/ffmpeg_memory
